start github-and-away tech note (#1167)
ropensci/roweb3: content/blog/2025-12-17-github-and-away/index.md

Change events originating on the primary remote are generally incorporated in a local version via `git pull`, and then pushed out to all mirrored versions.
A `git pull` command should only ever be applied to the primary remote version, and never to any alternative mirror versions.
In this diagram, the large yellow arrow between the primary remote and local represents the only connection where both `push` and `pull` events are allowed.
All other arrows are `push` events only, and code is never `pull`ed from non-primary mirrors.

{{< figure src = "local-remote.png" alt = "" class = "pull-left" caption = "Git interaction directions between local and remote repositories.">}}

### Mirroring versus Migration

The term "mirroring" is used here, and [throughout the official git documentation](https://git-scm.com/search/results?search=mirror&language=en), to imply a duplication of code, generally from local to remote locations.
"Migration" implies moving code away from one location to a new location.
Migration is neither necessary nor encouraged here, yet some sites nevertheless use this term, presumably as an implicit encouragement to do exactly that.

> Wherever "migration" is used below, please just think of a process of "mirroring" in which you code will always remain intact in the original location too.

Reviewer: I'd move this paragraph to come right after "Code Mirroring"

{{< figure src = "codeberg-new-migration.png" alt = "" class = "pull-left" caption = "Codeberg new repository migration button.">}}

That will then open up the following grid of options from where you want to migrate your repository:

{{< figure src = "codeberg-migration-options.png" alt = "Codeberg repository migration options shown as icons including Git, GitHub, GitLab, Forgejo, Gitea, Gogs, OneDev, and GitBucket" class = "pull-left" caption = "Codeberg repository migration options.">}}```

To migrate from GitHub, click the symbol to open a migration to fill in some details, where you can also paste a GitHub token into "access token" (ensuring that it has at least read access to all aspects you want transferred), and migrate almost all other aspects, including issues, pull requests, and releases.
Note that the migration process may take 10 minutes or more.

On GitLab, the equivalent "+" button is on the top left, where "_New project/repository_" should be clicked.

{{< figure src = "gitlab-new.png" alt = "" class = "pull-left" caption = "GitLab new repository button.">}}

> That will lead to the following options, in which "Import project" should be clicked.

Reviewer: ```suggestion
That will lead to a series of options, in which "Import project" should be clicked.
```

All other hyperlinks, including any to GitHub users, or to other repositories, will be lost.
The transfer will be of information at the time the migration or import was initiated.
From that time point on, any activity via issues or pull requests on any hosting platform will be specific to that platform only.[^2]

[^2]: There are tools for continuous mirroring such as [GitHubCodebergMirror](https://codeberg.org/wl/GithubCodebergMirror), but there is likely some fragility in these setups. An alternative is the approach adopted by the [zig language on Codeberg](https://ziglang.org/news/migrating-from-github-to-codeberg#:~:text=start%20counting%20issues) of simply ensuring the new primary mirror has unambiguous issue numbering.

### Migrating elsewhere

No other platforms currently offer the one-click migration functionality of Codeberg or GitLab.
To migrate in all other cases, you'll need to:

1. Create a new repository on the desired platform.
2. Set a `git remote` URL to the new destination.
3. `git push` to new remote.

The [git remote web page](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes) provides more detail on working with remotes.

> We are also not currently aware of any platforms other than Codeberg and GitLab which offer inbuilt ability to import additional information on issues and pull requests, although other people have implemented things such as [issue mirroring from GitHub to sourcehut](https://github.com/marketplace/actions/sourcehut-issue-maker).

Reviewer: ```suggestion
We are also not currently aware of any platforms other than Codeberg and GitLab which offer inbuilt ability to import additional information on issues and pull requests, although other people have implemented things such as [issue mirroring from GitHub to SourceHut](https://github.com/marketplace/actions/sourcehut-issue-maker).
```

Other code hosting platforms like [SourceHut](https://sr.ht/), or the distributed hosting system [tangled](https://tangled.org/) do not provide facilities for directly transferring or mirroring issues or pull requests (and in fact generally don't even have such things, as they really are specific to the GitHub platform; [SourceHut](https://sr.ht), for example, implements its own systems of "todos" and "patches").

## Mirroring: Managing one repository across multiple platforms

> The goal here is to move from having code hosted on a single platform to distributing code across multiple platforms through _mirroring_. This reduces the risk of dependence on a single platform creating a single point of failure. Mirroring also allows you to use alternative platforms to GitHub, while will benefiting from GitHub operations used by other organizations (like rOpenSci!). To achieve this goal, you'll first need to add the new remote locations to your local git configuration, and then synchronize those remotes.

Reviewer: ```suggestion
The goal here is to move from having code hosted on a single platform to distributing code across multiple platforms through _mirroring_. This reduces the risk of dependence on a single platform creating a single point of failure. Mirroring also allows you to use alternative platforms to GitHub, while still benefiting from GitHub operations used by other organizations (like rOpenSci!). To achieve this goal, after setting up your mirrors you'll next need to add the new remote locations to your local git configuration, and then synchronize those remotes.
```
